crates/wash-lib/src/cli/stop.rs

```diff
@@ -217,3 +265,58 @@ pub async fn stop_host(cmd: StopHostCommand) -> Result<CommandOutput> {
         format!("Host {} acknowledged stop request", cmd.host_id),
     ))
 }
+
+async fn find_host_with_provider(
+    provider_id: &str,
+    ctl_client: &wasmcloud_control_interface::Client,
+) -> Result<ServerId, FindIdError> {
+    find_host_with_filter(ctl_client, |inv| {
+        inv.providers
+            .into_iter()
+            .any(|prov| prov.id == provider_id)
+            .then_some((inv.host_id, inv.friendly_name))
+            .and_then(|(id, friendly_name)| id.parse().ok().map(|i| (i, friendly_name)))
+    })
+    .await
+}
+
+pub(crate) async fn find_host_with_actor(
+    actor_id: &str,
+    ctl_client: &wasmcloud_control_interface::Client,
+) -> Result<ServerId, FindIdError> {
+    find_host_with_filter(ctl_client, |inv| {
+        inv.actors
+            .into_iter()
+            .any(|actor| actor.id == actor_id)
+            .then_some((inv.host_id, inv.friendly_name))
+            .and_then(|(id, friendly_name)| id.parse().ok().map(|i| (i, friendly_name)))
+    })
+    .await
+}
+
+async fn find_host_with_filter<F>(
+    ctl_client: &wasmcloud_control_interface::Client,
+    filter: F,
+) -> Result<ServerId, FindIdError>
+where
+    F: FnMut(HostInventory) -> Option<(ServerId, String)>,
+{
+    let inventories = get_all_inventories(ctl_client).await?;
+    let all_matching = inventories
+        .into_iter()
+        .filter_map(filter)
+        .collect::<Vec<(ServerId, String)>>();
+
+    if all_matching.is_empty() {
+        Err(FindIdError::NoMatches)
+    } else if all_matching.len() > 1 {
+        let matches = all_matching.into_iter().map(|(id, friendly_name)| Match {
+            id: id.into_string(),
+            friendly_name: Some(friendly_name),
+        });
+        Err(FindIdError::MultipleMatches(matches.collect()))
+    } else {
+        // SAFETY: We know there is exactly one match at this point
+        Ok(all_matching.into_iter().next().unwrap().0)
+    }
```

Comment:
How do you feel about using `match all_matching[..]` ? A bit easier to read IMO and then you don't need the safety comment I think (it's self-evident).

Also intermediate `matches` var probably unnecessary

Reply:
I did split up `matches` on purpose to make it a bit more readable, but I don't care either way how we do it. Let me try the match pattern though

Reply:
Ok, so I tried this and it doesn't like matching the whole slice (and I am pretty sure it borrows rather than taking ownership, which is an admittedly minor detail here), so I'm not going to fight with it right now

Reply:
Thanks for trying it, definitely not urgent so all good.